README.md
## Neo4j.Sips

A simple Elixir driver using the [Neo4j](http://neo4j.com/developer/get-started/) graph database REST API. Compatible with the following Neo4j servers: `2.x/3.0.x/3.1.x`

![Build Status](https://travis-ci.org/florinpatrascu/neo4j_sips.svg?branch=master)
[![Deps Status](https://beta.hexfaktor.org/badge/all/github/florinpatrascu/neo4j_sips.svg)](https://beta.hexfaktor.org/github/florinpatrascu/neo4j_sips)
[![Hex.pm](https://img.shields.io/hexpm/dt/neo4j_sips.svg?maxAge=2592000)](https://hex.pm/packages/neo4j_sips)
[![Hexdocs.pm](https://img.shields.io/badge/api-hexdocs-brightgreen.svg)](https://hexdocs.pm/neo4j_sips)

Documentation: [hexdocs.pm/neo4j_sips/](http://hexdocs.pm/neo4j_sips/)

*You can also look at: [Bolt.Sips](https://github.com/florinpatrascu/bolt_sips) - Elixir driver using the Bolt protocol; Neo4j's newest network protocol, designed for high-performance.*

### Install

[Available in Hex](https://hex.pm/packages/neo4j_sips). Edit the `mix.ex` file and add the `neo4j_sips` dependency to the `deps/1 `function:

    def deps do
      [{:neo4j_sips, "~> 0.2"}]
    end

or from Github:

    def deps do
      [{:neo4j_sips, github: "florinpatrascu/neo4j_sips"}]
    end

If you're using a local development copy:

    def deps do
      [{:neo4j_sips, path: "../neo4j_sips"}]
    end

Then add the `neo4j_sips` dependency the applications list:

    def application do
      [applications: [:logger, :neo4j_sips],
       mod: {Neo4j.Sips.Application, []}]
    end


Edit the `config/config.exs` and describe a Neo4j server endpoint, example:

    config :neo4j_sips, Neo4j,
      url: "http://localhost:7474",
      pool_size: 5,
      max_overflow: 2,
      timeout: 30

Run `mix do deps.get, deps.compile`

If your server requires basic authentication, add this to your config file:
      
      basic_auth: [username: "foo", password: "bar"]
      
Or:
      
      token_auth: "bmVvNGo6dGVzdA==" # if using an authentication token?!

You can also specify the authentication in the `url` config:

      url: "http://neo4j:neo4j@localhost:7474"
  
### Example

With a minimalist setup configured as above, and a Neo4j server running, we can connect to the server and run some queries using Elixir’s interactive shell ([IEx](http://elixir-lang.org/docs/stable/iex/IEx.html)):

    $ cd <my_mix_project>
    $ iex -S mix
    Erlang/OTP 19 [erts-8.0.2] [source] [64-bit] ...

    Interactive Elixir (1.3.2) - press Ctrl+C to exit (type h() ENTER for help)
    iex(1)> alias Neo4j.Sips, as: Neo4j

    iex(2)> Neo4j.start_link(url: "http://localhost:7474")
    {:ok, #PID<0.204.0>}

    iex(3)> cypher = """
      CREATE (n:Neo4jSips {title:'Elixir sipping from Neo4j', released:2015, 
        license:'MIT', neo4j_sips_test: true})
    """

    iex(4)> Neo4j.query(Neo4j.conn, cypher)
    {:ok, []}

    iex(5)> n = Neo4j.query!(Neo4j.conn, "match (n:Neo4jSips {title:'Elixir sipping from Neo4j'}) where n.neo4j_sips_test return n")
    [%{"n" => %{"license" => "MIT", "neo4j_sips_test" => true, "released" => 2015,
         "title" => "Elixir sipping from Neo4j"}}]
      
For more examples, see the test suites.

### Contributing

- [Fork it](https://github.com/florinpatrascu/neo4j_sips/fork)
- Create your feature branch (`git checkout -b my-new-feature`)
- Test (`mix test`)
- Commit your changes (`git commit -am 'Add some feature'`)
- Push to the branch (`git push origin my-new-feature`)
- Create new Pull Request

### Contributors

As reported by Github: [contributions to master, excluding merge commits](https://github.com/florinpatrascu/neo4j_sips/graphs/contributors)

### Author
Florin T.PATRASCU (@florinpatrascu, @florin on Twitter)

## License
* Neo4j.Sips - MIT, check [LICENSE](LICENSE) file for more information.
* Neo4j - Dual free software/commercial license, see http://neo4j.org/


CHANGELOG.md
# Changelog

## v0.2.18

(via https://github.com/florinpatrascu/neo4j_sips/pull/21, thank you: https://github.com/suddenrushofsushi) Currently setting :timeout in Neo4j.Sips config sets the timeout for :poolboy.transaction but neither HTTP calls, nor the :gen_server call wrapping them have the timeout applied.

This version adds:

- HTTP connection and receive timeouts to HTTPoison calls
- Timeout to the gen_server call which wraps the HTTP calls above.

## v0.2.17 (2017-01-09)
- relax the `:poison` library version requirements; in relation to Phoenix 1.2.x

## v0.2.16
- Travis test builds with Elixir 1.3/1.4
- tame some of the warnings and mild code refactoring
- updated README

## v0.2.15 (2017-01-05)
- add Travis CI support
- temporarily suspending the `Neo4j.Sips.Server.Test` suite; requires more thinking, after the url refactoring
- Elixir 1.3
- return an error if the driver authentication fails

## v0.2.14 (2016-12-26)
- Neo4j 3.1 is now returning the address of the `Bolt` protocol address, during the initial handshake, with the remote http API. At this time, I am expecting a set of keys I convert later to atoms, for efficiency. However, the story with the atoms in Erlang is well-known: `Atoms are not garbage-collected. Once an atom is created, it will never be removed.` This is why I also had to make sure I am allocating all the keys I need **before** this initial handshake. And the `:bolt` atom was not one of them, as I didn't expect to have it, breaking this way the Poison validations. Fixed now.

## v0.2.12 (2016-11-07)
- minor changes: dependencies update

## v0.2.11 (2016-09-29)
- fix access error bug. PR provided by @tpunt. Thank you.

## v0.2.10 (2016-07-24)
- Enhancements
    * ready for Elixir 1.3
    * Neo4j.Sips is paving the path towards an easier integration with third party frameworks i.e. where you may want to use Neo4j.Sips as an Ecto-like Repo, Adapters, etc.
    * you can enable the logger now and see the requests we send to the Neo4j server. Please see `config/test.exs`, for an example of logger configuration. For now the logged info is very simple simple, yet useful for debugging
    * added more tests
    * code cleanup and various code optimizations
- Breaking changes
    * you must start the `Neo4j.Sips` server process. This is easily done via: `Neo4j.Sips.start_link/1`. For example: `Neo4j.Sips.start_link(url: "http://localhost:7474")`
- Bug fixes
    * the driver configuration is properly reloaded


lib/neo4j_sips.ex
defmodule Neo4j.Sips do
  @moduledoc """
  Elixir driver for Neo4j

  A module that provides a simple Interface to communicate with a Neo4j server via http,
  using [Neo4j's own REST API](http://neo4j.com/docs/stable/rest-api.html).

  All functions take a pool to run the query on.
  """
  use Supervisor

  alias Neo4j.Sips.Transaction
  alias Neo4j.Sips.Connection
  alias Neo4j.Sips.Server
  alias Neo4j.Sips.Query
  alias Neo4j.Sips.Utils

  @pool_name :neo4j_sips_pool

  @doc """
  Example of valid configurations (i.e. defined in config/dev.exs):

     # Neo4j server not requiring authentication
     config :neo4j_sips, Neo4j,
       url: "http://localhost:7474"

     # Neo4j server with username and password authentication
     config :neo4j_sips, Neo4j,
       url: "http://localhost:7474",
       pool_size: 5,
       max_overflow: 2,
       timeout: 30,
       basic_auth: [username: "neo4j", password: "neo4j"]

     # or using a token
     config :neo4j_sips, Neo4j,
       url: "http://localhost:7474",
       pool_size: 10,
       max_overflow: 5,
       timeout: :infinity,
       token_auth: "bmVvNGo6dGVzdA=="
  """
  def start_link(opts) do
    cnf = Utils.default_config(opts)

    ConCache.start_link([], name: :neo4j_sips_cache)
    ConCache.put(:neo4j_sips_cache, :config, cnf)

    poolboy_config = [
      name: {:local, @pool_name},
      worker_module: Neo4j.Sips.Connection,
      size: Keyword.get(cnf, :pool_size),
      max_overflow: Keyword.get(cnf, :max_overflow)
    ]

    case Server.init(cnf) do
      {:ok, server} ->
        ConCache.put(:neo4j_sips_cache, :conn, %Neo4j.Sips.Connection{
                    server: server,
                    transaction_url: server.data.transaction,
                    server_version: server.data.neo4j_version,
                    commit_url: "",
                    options: nil
                  })

      {:error, message} -> Mix.raise message
    end

    children = [:poolboy.child_spec(@pool_name, poolboy_config, cnf)]
    options = [strategy: :one_for_one, name: __MODULE__]

    Supervisor.start_link(children, options)
  end

  @doc false
  def child_spec(opts) do
    Supervisor.Spec.worker(__MODULE__, [opts])
  end

  ## Connection

  @doc """
  returns a Connection containing the server details. You can
  specify some optional parameters i.e. graph_result.

  graph_result is nil, by default, and can have the following values:

      graph_result: ["row"]
      graph_result: ["graph"]
  or both:

      graph_result: [ "row", "graph" ]

  """
  defdelegate conn(options), to: Connection

  # until defdelegate allows optional args?!
  @doc """
  returns a Neo4j.Sips.Connection
  """
  defdelegate conn(), to: Connection

  @doc """

  returns the server version
  """
  @spec server_version() :: String.t
  defdelegate server_version(), to: Connection

  ## Query
  ########################

  @doc """
  sends the query (and its parameters) to the server and returns `{:ok, Neo4j.Sips.Response}` or
  `{:error, error}` otherwise
  """
  @spec query(Neo4j.Sips.Connection, String.t) :: {:ok, Neo4j.Sips.Response} | {:error, Neo4j.Sips.Error}
  defdelegate query(conn, statement), to: Query

  @doc """
  The same as query/2 but raises a Neo4j.Sips.Error if it fails.
  Returns the server response otherwise.
  """
  @spec query!(Neo4j.Sips.Connection, String.t) :: Neo4j.Sips.Response | Neo4j.Sips.Error
  defdelegate query!(conn, statement), to: Query

  @doc """
  send a query and an associated map of parameters. Returns the server response or an error
  """
  @spec query(Neo4j.Sips.Connection, String.t, Map.t) :: {:ok, Neo4j.Sips.Response} | {:error, Neo4j.Sips.Error}
  defdelegate query(conn, statement, params), to: Query

  @doc """
  The same as query/3 but raises a Neo4j.Sips.Error if it fails.
  """
  @spec query!(Neo4j.Sips.Connection, String.t, Map.t) :: Neo4j.Sips.Response | Neo4j.Sips.Error
  defdelegate query!(conn, statement, params), to: Query


  ## Transaction
  ########################

  @doc """
  begin a new transaction.
  """
  @spec tx_begin(Neo4j.Sips.Connection) :: Neo4j.Sips.Connection
  defdelegate tx_begin(conn), to: Transaction

  @doc """
  execute a Cypher statement in a new or an existing transaction
  begin a new transaction. If there is no need to keep a
  transaction open across multiple HTTP requests, you can begin a transaction,
  execute statements, and commit with just a single HTTP request.
  """
  @spec tx_commit(Neo4j.Sips.Connection, String.t) :: Neo4j.Sips.Response
  defdelegate tx_commit(conn, statements), to: Transaction

  @doc """
  given you have an open transaction, you can use this to send a commit request
  """
  @spec tx_commit(Neo4j.Sips.Connection) :: Neo4j.Sips.Response
  defdelegate tx_commit(conn), to: Transaction

  @doc """
  execute a Cypher statement with a map containing associated parameters
  """
  @spec tx_commit(Neo4j.Sips.Connection, String.t, Map.t) :: Neo4j.Sips.Response
  defdelegate tx_commit(conn, statement, params), to: Transaction

  @spec tx_commit!(Neo4j.Sips.Connection, String.t) :: Neo4j.Sips.Response
  defdelegate tx_commit!(conn, statements), to: Transaction

  @spec tx_commit!(Neo4j.Sips.Connection, String.t, Map.t) :: Neo4j.Sips.Response
  defdelegate tx_commit!(conn, statement, params), to: Transaction

  @doc """
  given that you have an open transaction, you can send a rollback request.
  The server will rollback the transaction. Any further statements trying to run
  in this transaction will fail immediately.
  """
  @spec tx_rollback(Neo4j.Sips.Connection) :: Neo4j.Sips.Connection
  defdelegate tx_rollback(conn), to: Transaction

  @doc """
  list all property keys ever used in the database. This also includes any property
  keys you have used, but deleted. There is currently no way to tell which ones
  are in use and which ones are not, short of walking the entire set of properties
  in the database.
  """
  @spec property_keys() :: List.t | []
  def property_keys do
    property_keys_url = Neo4j.Sips.conn.server.data_url <> "propertykeys"
    Connection.send(:get, property_keys_url)
  end

  @doc """
   returns an environment specific Neo4j.Sips configuration.
  """
  def config, do: ConCache.get(:neo4j_sips_cache, :config)

  @doc false
  def config(key), do: Keyword.get(config(), key)

  @doc false
  def config(key, default), do: Keyword.get(config(), key, default)

  @doc false
  def pool_name, do: @pool_name

  @doc false
  def init(args) do
    {:ok, args}
  end
end
